pkgs/servers/http/nginx/nix-etag-1.15.4.patch

```diff
@@ -0,0 +1,87 @@
+From 3ed81ef9e973247d0c95bf240b3bd2e640f2605a Mon Sep 17 00:00:00 2001
+From: Yegor Timoshenko <yegortimoshenko@riseup.net>
+Date: Fri, 28 Sep 2018 03:27:04 +0000
+Subject: [PATCH] If root is in Nix store, set ETag to its path hash
+
+---
+ src/http/ngx_http_core_module.c | 56 +++++++++++++++++++++++++++------
+ 1 file changed, 47 insertions(+), 9 deletions(-)
+
+diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
+index c57ec00c..d996d7e3 100644
+--- a/src/http/ngx_http_core_module.c
++++ b/src/http/ngx_http_core_module.c
+@@ -1583,6 +1583,7 @@ ngx_http_set_etag(ngx_http_request_t *r)
+ {
+     ngx_table_elt_t           *etag;
+     ngx_http_core_loc_conf_t  *clcf;
++    u_char                    *real, *ptr1, *ptr2;
+ 
+     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
+ 
+@@ -1598,16 +1599,53 @@ ngx_http_set_etag(ngx_http_request_t *r)
+     etag->hash = 1;
+     ngx_str_set(&etag->key, "ETag");
+ 
+-    etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
+-    if (etag->value.data == NULL) {
+-        etag->hash = 0;
+-        return NGX_ERROR;
+-    }
++    real = ngx_realpath(clcf->root.data, NULL);
++
++    #define NIX_STORE_DIR "/nix/store"
++    #define NIX_STORE_LEN ngx_strlen(NIX_STORE_DIR)
++
++    if (!ngx_strncmp(real, NIX_STORE_DIR, NIX_STORE_LEN)
```

Comment:
Does this segfault if `ngx_realpath` fails and hence `real` is null?

Reply:
It's undefined behavior. Thanks for the catch! It won't segfault on GCC, but will on Clang.

Reply:
Oh, it segfaults very well.  It just doesn't happen in your example since `strncmp` is a pure function.  The compiler then [optimizes it away](https://godbolt.org/z/oBa3bT) because the result is not used.
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STRING "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz"

int main() {
  // segfault
  printf("%i\n", strncmp(NULL, STRING, strlen(STRING)));
}
```

Reply:
Right, I've edited post above :-) Also, fixed in https://raw.githubusercontent.com/NixOS/nixpkgs/971f8e2c779f57a3dfc931161d266b2c32b1d67e/pkgs/servers/http/nginx/nix-etag-1.15.4.patch.